providers/implementations/rands/drbg.c

```diff
@@ -112,23 +112,13 @@ static unsigned int get_parent_reseed_count(PROV_DRBG *drbg)
     void *parent = drbg->parent;
     unsigned int r;
 
-    *params = OSSL_PARAM_construct_uint(OSSL_DRBG_PARAM_RESEED_CTR, &r);
-    if (!drbg_lock_parent(drbg)) {
-        ERR_raise(ERR_LIB_PROV, PROV_R_UNABLE_TO_LOCK_PARENT);
-        goto err;
-    }
+    *params = OSSL_PARAM_construct_uint(OSSL_DRBG_PARAM_RESEED_COUNTER, &r);
+    /* we don't call drbg_lock_parent(drbg), because an atomic read is used */
```

Comment:
I somewhat disagree here.  The parent DRBG need not be any of our DRBGs.  This one is deliberately permitted to be cross the lib context boundary.  By assuming an atomic read, the semantics are weakened.  Using a lock (for which a call back is specially supplied) is the safest option.

The ideal is a an optional local and an atomic read and a big comment here saying why :)

Reply:
> The ideal is a an optional ~~local~~ lock and an atomic read and a big comment here saying why :)

Maybe you are right. Using atomics instead of a lock was a natural choice and a simple performance improvement back in 1.1.1.  _(Actually, the optimal compromise between safety and performance would have been a reader-writer lock, but since they are not available, I went with atomics instead of using full locks.)_ But even in 1.1.1 it turned out that it's much harder to get lock-free synchronization right.

Now with the 3.0 refactoring, the simple read of an integer has turned into a function call to possibly another provider. In this case, it's even harder to get it right and a lock seems to be the safer solution, even if it has a performance impact (which we didn't measure).

Reply:
Any performance impact will be related to the number of contented reseeds.  Longer chains, fewer contentions.

The last time I looked properly at the impact of RNG locking was for 1.0.2 and it did terrible things to connections per second.  We won't have that problem here.